src/Controller/ApiV1Controller.php

```diff
@@ -0,0 +1,211 @@
+<?php
+/**
+ * Tatoeba Project, free collaborative creation of multilingual corpuses project
+ * Copyright (C) 2009-2010  HO Ngoc Phuong Trang <tranglich@gmail.com>
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * PHP version 5
+ *
+ * @category PHP
+ * @package  Tatoeba
+ * @author   HO Ngoc Phuong Trang <tranglich@gmail.com>
+ * @license  Affero General Public License
+ * @link     https://tatoeba.org
+ */
+namespace App\Controller;
+
+use App\Controller\AppController;
+use App\Form\SentencesSearchForm;
+use App\Model\CurrentUser;
+use App\Model\Table\SentencesTable;
+use App\Lib\LanguagesLib;
+use App\Lib\SphinxClient;
+use App\Lib\Licenses;
+use Cake\Core\Configure;
+use Cake\Database\Expression\QueryExpression;
+use Cake\Event\Event;
+use Cake\Routing\Router;
+use Cake\Utility\Hash;
+use Cake\View\ViewBuilder;
+use Exception;
+
+/**
+ * Controller for Api.
+ *
+ * @category Sentences
+ * @package  Controllers
+ * @author   HO Ngoc Phuong Trang <tranglich@gmail.com>
+ * @license  Affero General Public License
+ * @link     https://tatoeba.org
+ */
+class ApiV1Controller extends AppController
+{
+    public $name = 'Sentences';
+    public $components = array (
+        'LanguageDetection',
+        'CommonSentence',
+        'Permissions',
+    );
+    public $helpers = array(
+        'Sentences',
+        'Menu',
+        'Lists',
+        'SentenceButtons',
+        'Html',
+        'Logs',
+        'Pagination',
+        'Comments',
+        'Languages',
+        'CommonModules'
+    );
+
+    public $uses = array(
+        'Audio',
+        'Sentence',
+        'SentencesSentencesLists',
+        'SentencesList',
+        'User',
+        'UsersLanguages',
+        'Tag',
+        'UsersSentences',
+        'Vocabulary'
+    );
+
+    public function initialize() {
+        parent::initialize();
+
+        $params = $this->request->params;
+        $noCsrfActions = ['edit_sentence', 'change_language'];
+        if (in_array($params['action'], $noCsrfActions)) {
+            $this->components()->unload('Csrf');
+        }
+    }
+
+    /**
+     * Show Json of specified sentence id.
+     * 
+     * @param mixed $id Id of the sentence.
+     * 
+     * @return void
+     */
+    public function sentence($id = null) {
```

Comment:
By writing ` = null`, you are defining the `$id` parameter as optional. This will make the `/api_v1/sentence` a valid URL that will call this method with `$id` set to null. However, I think you want to enforce this parameter instead.

Reply:
OK

Reply:
Yes. As scary as it may look, I think the "too few arguments" error is what we want here because `/api_v1/sentence/` is not a valid URI. Since we are building an API, the point is reply with an appropriate [HTTP status code](https://en.wikipedia.org/wiki/List_of_HTTP_status_codes), and we don’t care about how it looks in the browser. Basically you should return the status 200 if there is a sentence, otherwise some 4xx or 5xx error. One way to check the actual return code is to use the "Developer Tools" of your browser (in the "Network" tab).